extensions/amp-sidebar/1.0/sidebar.jss.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {createUseStyles} from 'react-jss';
+
+const baseClass = {
+  position: 'fixed !important',
+  top: 0,
+  maxHeight: '100vh !important',
+  height: '100vh',
+  maxWidth: '80vw',
+  backgroundColor: '#efefef',
+  minWidth: '45px !important',
+  outline: 'none',
+  overflowX: 'hidden !important',
+  overflowY: 'auto !important',
+  zIndex: 2147483647,
+};
+
+const left = {
+  left: 0,
+  transform: 'translateX(-100%)',
```

Comment:
Can we avoid `transform` here and below? It's best to define the stylesheets with the final values and leave offset values to the animation.

Reply:
This is a good point and I also considered this.  What I noticed is that since the order is Render with original css (from jss here).  Then do the animation.  I noticed that if we remove this line, Preact will first render the component for an instant, then start the animation.  So in this case, it will flash the sidebar fully opened, then start animating it from the side.  Same with opacity below and the mask.

Reply:
Sure, that definitely should be taken care of. There are a couple of solutions here. But I think in this case the easiest for us is:

1. Switch to `useLayoutEffect()`
2. Set the initial pre-animation state before animation is started.

E.g.

```
useLayoutEffect(() => {
  ...
  if (opened) {
    ...
    ...
    setStyles(sidebarElement, SLIDE_FROM_LEFT_INI_STYLES);
    sidebarAnimation = sidebarElement.animate(SLIDE_FROM_LEFT_ANIM);
  }
}, ...);
```

Typically none of this is need in the closing animations - only opening has this "initialization" problem.

Why this is better? Because right in that scope you're also setting final styles, such as `setStyle(sidebarElement, 'transform', 'translateX(0)');`. Thus it makes it very clear which styles are affected by the animation from start to end.

And, btw, if you set the ini style before animation to something like `transform: translateX(-100%)`, that would mean that the best thing is to reset it to empty string in the end instead of `translateX(0)` (which is equivalent to removing it from style). That's because the animation effect is temporary. And it's nice that it'd set itself while running and remove itself cleanly once done.

Reply:
Awesome, thanks for the thorough explanation.  I've updated in the latest revision and factored out the styles as you've given in the example so that they are module level constants.  If you have opinions on the re-factoring please let me know, it might be overkill since we only apply two styles in animation.